pybpodapi/bpod/emulator/emulator.py

```diff
@@ -0,0 +1,668 @@
+                        if output_channel_type in \
+                                (BNC_CODE, WIRE_CODE, PORTS_CODE):
+                            self._state.output[
+                                output_channel] = event_type
+        logger.debug('Events mirrored.')
+
+    def override_input_state(self, channel_number, channel_value):
+        self._state.input[channel_number] = channel_value
+        event_type = self._state.input_type[channel_number]
+        if event_type not in (UART_CODE, USB_CODE):
+            event_bnc_position = \
+                self._hardware.channels.events_positions.Event_BNC
+            output_usb_position = \
+                self._hardware.channels.events_positions.output_USB
+            channel_value_offset = 0
+            if channel_value > 0:
+                channel_value_offset = 1
+            new_event = event_bnc_position - 1 + \
+                2 * (channel_number - output_usb_position) - \
+                channel_value_offset
+            self._manual_override_events.put(new_event)
+
+    def serial_override(self, channel_number, channel_value):
+        self._state.output[channel_number] = channel_value
+
+    def add_manual_override_event(self, event):
+        hardware_channels = self._hardware.channels
+        if isinstance(event, int):
+            try:
+                hardware_channels.get_event_name(event)
+            except IndexError as e:
+                raise EmulatorError(str(e))
+            else:
+                event_code = event
+        else:
+            if event not in hardware_channels.event_names:
+                raise EmulatorError('unknown event name')
+            event_code = hardware_channels.event_names.index(event)
+        self._manual_override_events.put(event_code)
+
+    def log_state_machine_info(self):
+        self._assert_state_machine()
+        sma = self._state_machine
+        highest_used_global_counter = sma.global_counters.get_max_index_used()
+        highest_used_global_timer = sma.global_timers.get_max_index_used()
+        highest_used_global_condition = sma.conditions.get_max_index_used()
+        highest_used_global_counter = (
+            0
+            if highest_used_global_counter is None
+            else highest_used_global_counter + 1
+        )
+        highest_used_global_timer = (
+            0
+            if highest_used_global_timer is None
+            else highest_used_global_timer + 1
+        )
+        highest_used_global_condition = (
+            0
+            if highest_used_global_condition is None
+            else highest_used_global_condition + 1
+        )
+        # State timer matrix (state timer transitions)
+        state_timer_matrix = []
+        for i in range(sma.total_states_added):
+            state_timer_matrix += [
+                sma.total_states_added
+                if math.isnan(sma.state_timer_matrix[i])
+                else sma.state_timer_matrix[i]
+            ]
```

Comment:
Since you are appending a single value, whose value differ from the branch, I think it'll be clearer to refactor it as follows:
```python
timer = sma.state_timer_matrix[i]
if math.isnan(timer)
  timer = sma.total_states_added
state_timer_matrix.append(timer)
```

Also, if utilize numpy here, I think we can replace the whole for loop:
```python
state_timer_matrix = sma.state_timer_matrix[i].copy()
np.nan_to_num(state_timer_matrix, copy=False, nan=sma.total_states_added) # Or I think: state_timer_matrix[np.isnan(state_timer_matrix)] = sma.total_states_added)
```

Reply:
Same first part of the comment about the parts below that uses `if math.isnan(...)`